src/core_ocean/shared/mpas_ocn_mesh.F

```diff
@@ -0,0 +1,842 @@
+   ! on CPU, accelerator
+   type (meshType), public :: &
+      ocnMesh              ! default mesh
+
+   !----------------------------------------------------------------------------
+   !
+   ! Public member functions
+   !
+   !----------------------------------------------------------------------------
+
+   public ::                     &
+      ocn_meshCreate,            &
+      ocn_meshUpdateFields,      &
+      ocn_meshDestroy
+
+!*******************************************************************************
+
+contains
+
+!*******************************************************************************
+!
+!  ocn_meshCreate
+!
+!> \brief Creates the primary ocean mesh data structure on both host and device
+!!  \author Rob Aulwes and Phil Jones
+!!  \date   14 Jan 2020
+!!  \details
+!!  This module creates and maintains a primary ocean mesh structure
+!!  and ensures all mesh variables are copied to an accelerator device
+!!  if needed.
+!
+!-------------------------------------------------------------------------------
+
+   subroutine ocn_meshCreate(domain)
+
+      ! Input arguments
+
+      type(domain_type) :: &
+         domain                    !< [in] MPAS type to describe domain
+
+      !{{{
+      ! Local variables
+
+      integer :: &
+         blockCount               ! counter for number of blocks
+
+      type(block_type), pointer :: &
+         block                    ! variables in current subblock
+
+      type (mpas_pool_type), pointer :: &
+         meshPool                 ! mesh variables in MPAS pool structure
+
+      ! scalar pointers for retrieval, but convert to actual scalars in struct
+      integer, pointer ::   &! mesh dimensions
+         nCellsTmp,         &!
+         nEdgesTmp,         &!
+         nVerticesTmp,      &!
+         maxEdgesTmp,       &!
+         maxEdges2Tmp,      &!
+         vertexDegreeTmp,   &!
+         nVertLevelsTmp,    &!
+         nVertLevelsP1Tmp,  &!
+         nAdvectionCellsTmp  !
+
+      ! temporary pointers for converting masks
+      integer i,k,n          ! loop indices
+      integer, dimension(:,:), pointer :: &
+         edgeMaskTmp,                     &
+         vertexMaskTmp,                   &
+         cellMaskTmp,                     &
+         highOrderAdvectionMaskTmp,       &
+         edgeSignOnCellTmp,               &
+         edgeSignOnVertexTmp,             &
+         boundaryEdgeTmp,                 &
+         boundaryVertexTmp,               &
+         boundaryCellTmp
+
+      !***
+      !*** end of preamble, begin code
+      !***
+
+      !$omp master
+
+      blockCount = 0
+      block => domain % blocklist
+      do while ( associated(block) )
+
+         !*** we now only support one block per task
+         blockCount = blockCount + 1
+         if (blockCount > 1) then
+            call mpas_log_write(&
+               'ocn_meshCreate: more than one block no longer supported', &
+                MPAS_LOG_CRIT)
+         endif
```

Comment:
If this only supports 1 block per task, why is there the `do while` loop?  Is that the only way to get the number of blocks to confirm that there is only 1 block?

Reply:
You guessed it. The only way to get a block count (that I know of anyway).

Reply:
Well that's a bummer.  Although if we are getting rid of blocks soon then it won't be an issue for much longer.